libgadget/gravshort-tree.c

```diff
@@ -27,29 +26,71 @@
  *  short-range part.
  */
 
+static struct gravshort_tree_params TreeParams;
+
+/*This is a helper for the tests*/
+void set_gravshort_treepar(struct gravshort_tree_params tree_params)
+{
+    TreeParams = tree_params;
+}
+
+struct gravshort_tree_params get_gravshort_treepar(void)
+{
+    return TreeParams;
+}
+
+/* Sets up the module*/
+void
+set_gravshort_tree_params(ParameterSet * ps)
+{
+    int ThisTask;
+    MPI_Comm_rank(MPI_COMM_WORLD, &ThisTask);
+    if(ThisTask == 0) {
+        TreeParams.BHOpeningAngle = param_get_double(ps, "BHOpeningAngle");
+        TreeParams.ErrTolForceAcc = param_get_double(ps, "ErrTolForceAcc");
+        TreeParams.BHOpeningAngle = param_get_double(ps, "BHOpeningAngle");
+        TreeParams.TreeUseBH= param_get_int(ps, "TreeUseBH");
+        TreeParams.Rcut = param_get_double(ps, "TreeRcut");
+    }
+    MPI_Bcast(&TreeParams, sizeof(struct gravshort_tree_params), MPI_BYTE, 0, MPI_COMM_WORLD);
+}
+
 /* According to upstream P-GADGET3
  * correct workcount slows it down and yields little benefits in load balancing
  *
  * YF: anything we shall do about this?
  * */
 
-int force_treeev_shortrange(TreeWalkQueryGravShort * input,
+int
+force_treeev_shortrange(TreeWalkQueryGravShort * input,
         TreeWalkResultGravShort * output,
         LocalTreeWalk * lv);
 
+
 /*! This function computes the gravitational forces for all active particles.
  *  If needed, a new tree is constructed, otherwise the dynamically updated
  *  tree is used.  Particles are only exported to other processors when really
  *  needed, thereby allowing a good use of the communication buffer.
+ *  NeutrinoTracer = All.HybridNeutrinosOn && (All.Time <= All.HybridNuPartTime);
+ *  rho0 = All.CP.Omega0 * 3 * All.CP.Hubble * All.CP.Hubble / (8 * M_PI * All.G)
  */
-void grav_short_tree(ForceTree * tree)
+void
+grav_short_tree(const ActiveParticles * act, PetaPM * pm, ForceTree * tree, double G, double rho0, int NeutrinoTracer, int FastParticleType)
 {
     double timeall = 0;
     double timetree, timewait, timecomm;
-    if(!All.TreeGravOn)
-        return;
 
     TreeWalk tw[1] = {{0}};
+    struct GravShortPriv priv;
```

Comment:
I know this is OK because tw doesn't live outside of this functio.
It still feels there shall be an _init function and an _destroy function that maintain the life cycle of priv. (can probably move these initializers to the _init too?

Reply:
mm, it is overhead for only one use.